store/postgres/src/relational_queries.rs

```diff
@@ -803,6 +803,165 @@ impl<'a> QueryFragment<Pg> for PrefixComparison<'a> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct TableJoin<'a> {
+    parent: &'a Table,
+    child: &'a Table,
+    parent_prefix: String,
+    child_prefix: String,
+    parent_column: &'a Column,
+    child_column: &'a Column,
+}
+
+#[derive(Debug, Clone)]
+pub struct TableJoinCollection<'a> {
+    joins: Vec<TableJoin<'a>>,
+    prefixes: HashMap<String, String>,
+}
+
+impl<'a> TableJoinCollection<'a> {
+    pub fn new(root_prefix: &'a str, root_table: &'a Table) -> Self {
+        let mut prefixes = HashMap::new();
+
+        prefixes.insert(root_table.name.to_string(), root_prefix.to_string());
+
+        Self {
+            prefixes,
+            joins: Vec::new(),
+        }
+    }
+
+    pub fn extend_with_filter(
+        &mut self,
+        filter: &'a EntityFilter,
+        parent_table: &'a Table,
+        layout: &'a Layout,
+    ) -> Result<(), StoreError> {
+        use EntityFilter::*;
+        match filter {
+            And(filters) | Or(filters) => {
+                for filter in filters {
+                    self.extend_with_filter(filter, parent_table, layout)?;
+                }
+            }
+            Child(field_name, entity, child_filter) => {
+                let child_table = layout.table_for_entity(entity)?;
+
+                self.join(parent_table, child_table, field_name)?;
+                self.extend_with_filter(child_filter, child_table, layout)?;
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    fn find(&self, parent_table: &'a Table, child_table: &'a Table) -> Option<&TableJoin> {
+        self.joins
+            .iter()
+            .filter_map(|join| {
+                if join.parent.name == parent_table.name && join.child.name == child_table.name {
```

Comment:
This doesn't seem right - it's possible that the same parent and child tables appear in one join, e.g., if you have a schema like
```
type User {
  id ID!
}

type Account {
  id ID!
  owner User!
  sponsor User!
}
```

If you now filter accounts on both `owner` and `sponsor`, you'll have to join the account table with the user table twice for different purposes, and probably with different filters.

Reply:
It has now different, simpler implementation